File: java/socketd/src/main/java/org/noear/socketd/transport/core/StreamRequest.java
package org.noear.socketd.transport.core;

import org.noear.socketd.utils.IoConsumer;

/**
 * 请求流
 *
 * @author noear
 * @since 2.3
 */
public interface StreamRequest extends Stream<StreamRequest> {
    /**
     * 异步等待获取答复
     */
    Reply await();

    /**
     * 答复发生时
     */
    StreamRequest thenReply(IoConsumer<Reply> onReply);
}

File: python/socketd/socketd/transport/core/stream/StreamRequest.py
from socketd.transport.core.Message import Message
from socketd.transport.core.stream.StreamBase import StreamBase
from socketd.transport.utils.CompletableFuture import CompletableFuture


class StreamRequest(StreamBase):
    def __init__(self, sid, timeout, future):
        self.future: CompletableFuture = future
        super().__init__(sid, timeout)

    def is_single(self):
        return True

    def is_done(self):
        return self.future.done()

    def timeout(self):
        return self.timeout

    def on_accept(self, message: Message, onError):
        self.future.set_result(message)